### DISCUSSION TITLE
Is it possible to disable the global error only for specific requests of query, mutation?

### DISCUSSION THREAD

**Discussion by hoo00nn** [TanStack/query#3441] [Q&A] [answered]
Hello, Team react-query.

Thank you for providing a good library. I'm using it well.

However, I have some concerns while using it. Currently, I am handling errors for API requests globally in the QueryClient for query and mute requests.

Below are some of the codes I use.

``` typescript
new QueryClient({
    queryCache: new QueryCache({
      onError: (err) => errorHandler(err),
    }),
    mutationCache: new MutationCache({
      onError: (error, _, _) =>  errorHandler(error)
    }),
    defaultOptions: {
      queries: {
        refetchOnWindowFocus: false,
        retry: false,
      },
      mutations: {
        retry: false,
      },
    },
```

The problem I've experienced here is that I want to deal with errors customarily for specific requests.

When sending query, mutation requests, I expected that adding onError to options would be an override, but it was not.

Is there a way to handle the error I want instead of the global error in a particular request by passing it to the options in useMation?

I need your help. Please help me. 😭

**Comment by cyjo9603** [comment]
Oh, I'm having the same problem as you.  😭

**Comment by TkDodo** [comment, marked as answer]
> When sending query, mutation requests, I expected that adding onError to options would be an override, but it was not.

`onError` added to `useQuery` will overwrite the one passed to `defaultOptions`. However, those error handlers will be executed for every observer (=useQuery instance).

So I would stick to the global handler on the queryCache / mutationCache. Those handlers do get the whole query / mutation passed as the second parameter, so you can decide to not handle something for a specific key. Also, you can use the `meta` field on query/mutation to "tag" it in some way. This field is also available. For example, you could only show errors for queries that have a message passed:

```
const queryClient = new QueryClient({
    queryCache: new QueryCache({
      onError: (error, query) => {
        if (query.meta.message) {
         handleError(query.meta.message, error)
       }
      }
    }),
```

usage:

```
useQuery(key, fn, { meta: { message: "some message here" }})
```

this is just an example. `meta` can contain anything, so also boolean flags like `shouldBeHandledByGlobalErrorHandler: false` would do :)

**Reply by hoo00nn** [reply]
@TkDodo Thank you. I think I got a solution thanks to your reply. 😄 

I have one additional question. I would be very happy if I could get an answer.

I would like to manage objects for useMutation by domain. Examples are as follows.

``` typescript
export const useTodoMutation = () => {
  const cache = useQueryClient();

  const { mutate: createTodoMutation } = useMutation(createTodo, {
    onSuccess: async () => {
      await cache.invalidateQueries(queryKey.FETCH_TODO_LIST);
    },
  });

    const { mutate: deleteTodoMutation } = useMutation(deleteTodo, {
    onSuccess: async () => {
      await cache.invalidateQueries(queryKey.FETCH_TODO_LIST);
    },
  });

  const { mutate: updateTodoMutation } = useMutation(updateTodo, {
    onSuccess: async () => {
      await cache.invalidateQueries(queryKey.FETCH_TODO_LIST);
    },
  });

  return {
    createTodo: createTodoMutation,
    deleteTodo: deleteTodoMutation,
    updateTodo: updateTodoMutation,
  };
};
```

usage:

``` typescript
 import React, { FC } from 'react';

const TodoPage: FC = () => {
  const { createTodo } = useTodoMutation();

  const handleClickAddButton = async () => {
     await createTodo(todoItem);
  }

  return <></>;
};

export default TodoPage;
```

If you use it like above, I think you can use the code more neatly. Can't I use the meta field for mutate??

**Reply by TkDodo** [reply]
> Can't I use the meta field for mutate?

`meta` is a config option you have to pass to `useMutation`. It is currently not possible to pass it to `mutate`.

**Reply by hoo00nn** [reply]
@TkDodo  Thank you for your kind reply. 

Thank you again for creating a good library.

**Reply by palmmymumu** [reply]
@TkDodo  Seem like `meta` option isn't working with MutationCache unlike QueryCache ?

**Reply by TkDodo** [reply]
> @TkDodo Seem like meta option isn't working with MutationCache unlike QueryCache ?

what does "not working" mean? Please be more specific

**Reply by goacgras** [reply]
@TkDodo I try to use meta at my useMutation 
![image](https://github.com/TanStack/query/assets/66551792/0512c650-7fb7-4658-9340-f5964fa153ae)
nothing here 
![image](https://github.com/TanStack/query/assets/66551792/316852a7-5ad3-432e-9e38-12744488f59f)

Could you please explain how to use this meta, so that I can have some sortof tag at my global error handler

**Reply by TkDodo** [reply]
please look at the typescript types to see what the `onError` gets passed for the MutationCache:

https://github.com/TanStack/query/blob/39a4f05bc537cdd876d292336cb52dd8abf23a83/packages/query-core/src/mutationCache.ts#L14-L19

`meta` is in the `mutation`, so `mutation.meta`

**Comment by palmmymumu** [comment]
> > @TkDodo Seem like meta option isn't working with MutationCache unlike QueryCache ?
> 
> what does "not working" mean? Please be more specific

In MutationCache onSuccess or onError callback, on 2nd parameter `query` it doesn't have `meta` ?

**Reply by TkDodo** [reply]
the mutationCache callbacks don't get a Query passed (obviously). it's right there in the typescript types, so use typescript:

https://github.com/TanStack/query/blob/13b17ee45de3299dd3d3a2dfc3d23ad2f024882a/packages/query-core/src/mutationCache.ts#L14-L19

the parameters are

```
onError: (error, variables, context, mutation) => ...
```

so tl;dr: use the 4th parameter ...

**Comment by hshoja** [comment]
For the sake of convenience, we have come to this approach to define a distinct hook for handling specific exceptions.

```TS
export const DEFAULT_QUERY_CLIENT_CONFIG: Readonly<QueryClientConfig> = {
  defaultOptions: { queries: {... } },
  queryCache: new QueryCache({
    onError: (error, query) => {
      const meta = query?.meta ?? {}
      if ('preventGlobalError' in meta && meta.preventGlobalError) {
        reportError(error)
        return
      }
      throw error
    },
  }),
}

export const useNonThrowingQuery = <
  TQueryFnData = unknown,
  TError = DefaultError,
  TData = TQueryFnData,
  TQueryKey extends QueryKey = QueryKey,
>(
  options: UseQueryOptions<TQueryFnData, TError, TData, TQueryKey>,
  queryClient?: QueryClient,
): UseQueryResult<TData, TError> => {
  return useQuery<TQueryFnData, TError, TData, TQueryKey>(
    {
      meta: { preventGlobalError: true },
      ...options,
    },
    queryClient,
  )
}


```
